### PR TITLE
feat(curriculumn) added Run-length Encoding lesson

### PR DESCRIPTION
**Pull request by anthonyzeng99**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Pull Request contributing lessons discussed here:
https://github.com/freeCodeCamp/CurriculumExpansion/issues/296

### CODE REVIEW

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:5]: I just create a [guide article](https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-run-length-encoding/387031) on the forum for this challenge. Could you add `forumTopicId` to the frontmatter?:

````
---
id: 5eb257ba998e23057d209754
title: Run-length Encoding
challengeType: 5
forumTopicId: 387031
---
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:10]: We can reorganize this section a bit so it more closely resembles the original Rosetta Code description:

````
<section id='description'>
Given a string containing uppercase characters (A-Z), compress repeated 'runs' of the same character by storing the length of that run, and provide a function to reverse the compression.

Example:<br>
Original: `ABBBCDDDD`<br>
Encoded: `[[1, 'A'], [3, 'B'], [1, 'C'], [4, 'D']]`
</section>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:31]: Tried to play with the wording a bit here to make things just a little be clearer. Another thing we can do is adjust the spacing a bit here to take advantage of markdown-style code highlighting with backticks instead of `<code>` tags:

````
<section id='instructions'>

Write two functions, `encode` and `decode`.

`encode(input)` takes a string and returns the run-length encoding as a 2D array. Each element in the returned array is an array containing pairs, with the first element representing the number of occurrences and the second element representing the letter.

`decode(input)` takes a 2D array of pairs representing a run-length encoded string and returns the decoded string.

You may assume that the input will only contain uppercase letters.
</section>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:53]: Using the Before Test section to simplify the `testString`'s here is a great idea. But we should make each test case clear in the `text` for each of the tests:

````
tests:
  - text: <code>encode</code> should return an array.
    testString: assert(typeof encode(input_1) === 'object');
  - text: <code>decode</code> should return a string.
    testString: assert(typeof decode(output_1) === 'string');
  - text: <code>encode("ABBBCDDDD")</code> should return <code>[[1, 'A'], [3, 'B'], [1, 'C'], [4, 'D']]</code>
    testString: assert(arrayEquals(encode(input_1), output_1))
  - text: <code>decode([[1, 'A'], [3, 'B'], [1, 'C'], [4,'D']])</code> should return <code>"ABBBCDDDD"</code>
    testString: assert(decode(output_1) === input_1)
  - text: <code>encode("WWWWWWWWWWWWBWWWWWWWWWWWWBBBWWWWWWWWWWWWWWWWWWWWWWWWBWWWWWWWWWWWWWW")</code> should return <code>[[12, 'W'], [1, 'B'], [12, 'W'], [3, 'B'], [24, 'W'], [1, 'B'], [14, 'W']]</code>
    testString: assert(arrayEquals(encode(input_2), output_2))
  - text: <code>decode([[12, 'W'], [1, 'B'], [12, 'W'], [3, 'B'], [24, 'W'], [1, 'B'], [14, 'W']])</code> should return <code>"WWWWWWWWWWWWBWWWWWWWWWWWWBBBWWWWWWWWWWWWWWWWWWWWWWWWBWWWWWWWWWWWWWW"</code>
    testString: assert(decode(output_2) === input_2)
  - text: <code>encode("")</code> should return <code>[]</code>
    testString: assert(arrayEquals(encode(input_3), output_3))
  - text: <code>decode([])</code> should return <code>""</code>
    testString: assert(decode(output_3) === input_3)
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:69]: This is a minor detail, but maybe we could change the parameters here so it's a bit more obvious what we'll pass to each function behind the scenes:

```
function encode(str) {
  // Implement encode here
}
function decode(arr) {
  // Implement decode here
}
```

If you think that sounds good, we can adjust the Instructions section above to match.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:128]: Also, I got a linter error while running this locally. Please add a newline to the end of the file here.

**Comment by scissorsneedfoodtoo** [comment]
Hi @anthonyzeng99, thanks for all of your hard work here. 

I left some comments with some suggestions. Please take a look and let us know what you think.

**Comment by anthonyzeng99** [comment]
@scissorsneedfoodtoo Thanks for reviewing the PR. I have implemented the suggested changes.

**Comment by scissorsneedfoodtoo** [comment]
Awesome, thanks for taking care of that so quickly @anthonyzeng99.

Sorry, one more thing that I didn't notice last time. Please change the file name to `run-length-encoding.english.md `. Aside from the merge conflict, everything LGTM.

I tried to implement those changes myself and push to your branch, but unfortunately I don't have write access to nyu-ossd-s20.

Also, could you review this when you get a chance, @thecodingaviator?

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/run-length-encoding.md:14]:
```suggestion
Original: `ABBBAAZCDDDD`<br>
Encoded: `[[1, 'A'], [3, 'B'], [2, 'A'], [1, 'Z'], [1, 'C'], [4, 'D']]`
```
I'm suggesting this change to the example so that the user does not think that everything needs to be in alphabetical order. They can figure it out using the tests, but I'd rather add extra clarity.

**Review by thecodingaviator — CHANGES_REQUESTED** [review]

**Review by moT01 — CHANGES_REQUESTED** [review]
Can we also add the `isHidden: false` flag to this at the top.

**Comment by moT01** [comment]
I have blocked this for now. See the [pinned code freeze issue](https://github.com/freeCodeCamp/freeCodeCamp/issues/38308)

**Comment by raisedadead** [comment]
@anthonyzeng99 

Looks like you have explicitly disabled edits from maintainers, which means you would need to look into the currently updated file structure and rebase this PR before we are able to QA and merge this.

Thanks.

**Comment by moT01** [comment]
Ping @anthonyzeng99.

**Comment by RandellDawson** [comment]
Hi @anthonyzeng99 

As there has been no response from your end for a while, we are closing this PR.

Feel free to request this PR to be reopened or create a new PR, if you still wish to make the changes you originally added plus any additional change requests asked of you.  Also, remember to update your fork, so you have the latest version of master in case it has changed significantly.

Happy Coding!

**Pull request closed without merging**
